File: src/Tests/Behavioral/GenericTypeInference/GenericTypeInference.go
package main

import (
	"fmt"
)

type Signed interface {
	~int | ~int8 | ~int16 | ~int32 | ~int64
}

type Unsigned interface {
	~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr
}

type Integer interface {
	Signed | Unsigned
}

type Point []int32

func (p Point) String() string {
	return fmt.Sprintf("%d", p)
}

// Scale returns a copy of s with each element multiplied by c.
func Scale[S ~[]E, E Integer](s S, c E) S {
	r := make(S, len(s))
	for i, v := range s {
		r[i] = v * c
	}
	return r
}

// ScaleAndPrint doubles a Point and prints it.
func ScaleAndPrint(p Point) {
	r := Scale(p, 2)
	fmt.Println(r.String())
}

// Twice runs the constrained S back through ANOTHER generic function: Go infers the inner
// Scale's [S, E] through core types, but C# cannot re-infer a type parameter that appears only
// in constraints (CS0411 — the slices.Sort -> pdqsort chain), so the converter renders the
// resolved instantiation explicitly: `Scale<S, E>(s, 2)`.
func Twice[S ~[]E, E Integer](s S, c E) S {
	return Scale(s, c) // constrained S/E flow through: C# needs Scale<S, E>(s, c)
}

// CopyClearMinMax drives golib's interface-typed builtin overloads through the ~[]E constraint:
// copy with an ISlice-boxed dst must write into the caller's backing array, clear must zero it,
// and 2-arg min/max bind the IComparisonOperators overload (a constrained E has no IComparable).
func CopyClearMinMax[S ~[]E, E Integer](dst, src S) (E, E) {
	copy(dst, src)
	lo := min(dst[0], src[1])
	hi := max(dst[0], src[1])
	clear(src)
	return lo, hi
}

// SumHalves recurses over SUB-SLICES of the constrained S (the pdqsort shape): s[lo:hi] on a
// type parameter must yield S again SHARING backing (golib subslice<S, E> via S.Wrap), and
// append through the constraint must yield S (golib append<S, T>).
func SumHalves[S ~[]E, E Integer](s S) E {
	if len(s) == 1 {
		s[0] += s[0] // write through the sub-slice view -- must land in the caller's array
		return s[0]
	}
	mid := len(s) / 2
	return SumHalves(s[:mid]) + SumHalves(s[mid:])
}

// AppendKeep grows a constrained S through append, preserving its type.
func AppendKeep[S ~[]E, E Integer](s S, v E) S {
	out := append(s, v)
	return out
}

// setFirst takes a CONCRETE []E parameter.
func setFirst[E Integer](s []E, v E) {
	if len(s) > 0 {
		s[0] = v
	}
}

// PassSlice passes the constrained S where []E is expected (Go assignability -- the slices
// package's rotateRight/pdqsort helper chain): the emitted `new slice<E>(s)` materialization
// SHARES backing, so the helper's write lands in the caller's array.
func PassSlice[S ~[]E, E Integer](s S, v E) {
	setFirst(s, v)
}

// Grades is a named map type satisfying ~map[K]V.
type Grades map[string]int

// EqualMaps mirrors maps.Equal: comma-ok indexing through a ~map[K]V constrained M (the
// IMap<K, V> two-value indexer), with comparable-erased K/V equality routed through AreEqual.
func EqualMaps[M ~map[K]V, K, V comparable](m1, m2 M) bool {
	if len(m1) != len(m2) {
		return false
	}
	for k, v1 := range m1 {
		if v2, ok := m2[k]; !ok || v1 != v2 {
			return false
		}
	}
	return true
}

// Seq mirrors Go 1.23 iter.Seq: a GENERIC defined function type ranged over with
// range-over-func semantics (break must stop the producer via yield returning false).
type Seq[V any] func(yield func(V) bool)

// KVSeq mirrors iter.Seq2 (two-value range-over-func).
type KVSeq[K, V any] func(yield func(K, V) bool)

func countdown(n int) Seq[int] {
	return func(yield func(int) bool) {
		for i := n; i > 0; i-- {
			if !yield(i) {
				return
			}
		}
	}
}

func letters() KVSeq[string, int] {
	return func(yield func(string, int) bool) {
		_ = yield("a", 1) && yield("b", 2)
	}
}

// CloneOrNil mirrors maps.Clone: the nil-preserve guard (`m == nil` — Go's only legal map
// comparison, IMap.IsNil in C#) plus make/range/index-set through the constraint.
func CloneOrNil[M ~map[K]V, K, V comparable](m M) M {
	if m == nil {
		return nil
	}
	out := make(M, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

// DropKey mirrors maps.DeleteFunc's constrained delete.
func DropKey[M ~map[K]V, K, V comparable](m M, key K) {
	delete(m, key)
}

func main() {
	var p Point
	p = []int32{1, 2, 3}
	ScaleAndPrint(p)
	fmt.Println(Twice(Point{3, 4}, 2).String())
	dst2, src2 := Point{0, 0, 0}, Point{7, 8, 9}
	lo, hi := CopyClearMinMax(dst2, src2)
	fmt.Println(dst2.String(), src2.String(), lo, hi)
	q := Point{1, 2, 3, 4}
	total := SumHalves(q)
	fmt.Println(q.String(), total)
	grown := AppendKeep(Point{5, 6}, 7)
	fmt.Println(grown.String())
	pt := Point{1, 2}
	PassSlice(pt, 42)
	fmt.Println(pt.String())
	g1 := Grades{"a": 1, "b": 2}
	g2 := Grades{"a": 1, "b": 2}
	g3 := Grades{"a": 9}
	fmt.Println(EqualMaps(g1, g2), EqualMaps(g1, g3))
	sum := 0
	for v := range countdown(5) {
		if v == 2 {
			break // range-over-func break: yield returns false, producer stops
		}
		sum += v
	}
	fmt.Println(sum)
	for k, v := range letters() {
		fmt.Println(k, v)
	}
	var nilG Grades
	fmt.Println(CloneOrNil(nilG) == nil)
	cl := CloneOrNil(g1)
	fmt.Println(len(cl), cl["a"])
	DropKey(g1, "a")
	fmt.Println(len(g1))
	i16, ok16 := bsearchLike([]uint16{2, 4, 8}, uint16(4))
	i32, ok32 := bsearchLike([]uint32{10, 20, 30}, uint32(25))
	fmt.Println(i16, ok16, i32, ok32, halve(uint16(6)), halve(uint32(100)))
	fmt.Println(halveN(int32(10)), halveN(int64(-3)))
	ssum := int64(0)
	for v := range seqOf[int64](int32(4)) {
		ssum += v
	}
	fmt.Println(ssum)
	fmt.Println(growShrink(uintptr(1), uintptr(20)), growShrink(uint32(2), uint32(9)))
}

// growShrink instantiates a numeric-union constraint with U = uintptr: the golib uintptr
// STRUCT must DECLARE the generic-math interfaces the union maps to - matching operators
// alone never satisfy a where-clause (CS0315 x9 at reflect's rangeNum<uintptr, uint64>).
func growShrink[U ~uint32 | ~uintptr](v, lim U) U {
	for v < lim {
		v = v*2 + 1
	}
	return (v % lim) >> 1
}

// seqOf mirrors reflect's rangeNum[T, N]: TWO type parameters where the call supplies only T
// explicitly and N is inferred from the argument, returning a generic named type parameterized
// by T ALONE. The emitted C# call must carry the CALLEE's full resolved instantiation
// (info.Instances - both arguments), not the result Seq[T]'s single argument (CS0305 x11,
// reflect iter.go).
func seqOf[T ~int64, N ~int32 | ~int64](n N) Seq[T] {
	return func(yield func(T) bool) {
		for i := T(0); i < T(n); i = i + 1 {
			if !yield(i) {
				return
			}
		}
	}
}

// bsearchLike mirrors strconv's bsearch: a ~uint16|~uint32 union whose lifted Integer set
// must use the BCL shift shape IShiftOperators<E, int, E> (ushort/uint lack the self-typed form).
func bsearchLike[S ~[]E, E ~uint16 | ~uint32](s S, v E) (int, bool) {
	n := len(s)
	i, j := 0, n
	for i < j {
		h := i + (j-i)>>1
		if s[h] < v {
			i = h + 1
		} else {
			j = h
		}
	}
	return i, i < n && s[i] == v
}

// halveN mirrors rand.N[Int intType]: an integer-constrained type parameter compared
// against an untyped constant (CS0019 - the lifted operator interfaces take Int, not int)
// and converted to/from a basic integer (CS0030 - no cast exists to/from a type parameter;
// routed through golib ConvertToType/ConvertToUInt64, the banked E(100) family).
func halveN[Int ~int32 | ~int64](n Int) Int {
	if n <= 0 {
		return n
	}
	return Int(uint64(n) / 2)
}

// halve exercises a shift ON the type parameter itself: emits `v >> (int)(1)`,
// which binds IShiftOperators<E, int, E>.
func halve[E ~uint16 | ~uint32](v E) E {
	return v >> 1
}

File: src/Tests/Behavioral/GenericTypeInference/GenericTypeInference.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType("operators = Sum, Arithmetic, Integer, Comparable, Ordered")]
partial interface Signed<ΔT> {
    //  Type constraints: ~int | ~int8 | ~int16 | ~int32 | ~int64
    // Derived operators: +, -, *, /, %, &, |, ^, <<, >>, ==, !=, <, <=, >, >=
}

[GoType("operators = Sum, Arithmetic, Integer, Comparable, Ordered")]
partial interface Unsigned<ΔT> {
    //  Type constraints: ~uint | ~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uintptr
    // Derived operators: +, -, *, /, %, &, |, ^, <<, >>, ==, !=, <, <=, >, >=
}

[GoType("operators = Sum, Arithmetic, Integer, Comparable, Ordered")]
partial interface Integer<ΔT> {
    //  Type constraints: Signed | Unsigned
    // Derived operators: +, -, *, /, %, &, |, ^, <<, >>, ==, !=, <, <=, >, >=
}

[GoType("[]int32")] partial struct Point;

public static @string String(this Point p) {
    return fmt.Sprintf("%d"u8, p);
}

public static S Scale<S, E>(S s, E c)
    where S : /* ~[]E */ ISlice<E>, ISupportMake<S>, ISliceWrap<S, E>, new()
    where E : /* Integer */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    var r = make<S>(len(s));
    foreach (var (i, v) in s) {
        r[i] = v * c;
    }
    return r;
}

public static void ScaleAndPrint(Point p) {
    var r = Scale(p, 2);
    fmt.Println(r.String());
}

public static S Twice<S, E>(S s, E c)
    where S : /* ~[]E */ ISlice<E>, ISupportMake<S>, ISliceWrap<S, E>, new()
    where E : /* Integer */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    return Scale(s, c);
}

public static (E, E) CopyClearMinMax<S, E>(S dst, S src)
    where S : /* ~[]E */ ISlice<E>, ISupportMake<S>, ISliceWrap<S, E>, new()
    where E : /* Integer */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    copy(dst, src);
    var lo = min(dst[0], src[1]);
    var hi = max(dst[0], src[1]);
    clear(src);
    return (lo, hi);
}

public static E SumHalves<S, E>(S s)
    where S : /* ~[]E */ ISlice<E>, ISupportMake<S>, ISliceWrap<S, E>, new()
    where E : /* Integer */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    if (len(s) == 1) {
        s[0] += s[0];
        return s[0];
    }
    nint mid = len(s) / 2;
    return SumHalves<S, E>(subslice<S, E>(s, -1, mid)) + SumHalves<S, E>(subslice<S, E>(s, mid, -1));
}

public static S AppendKeep<S, E>(S s, E v)
    where S : /* ~[]E */ ISlice<E>, ISupportMake<S>, ISliceWrap<S, E>, new()
    where E : /* Integer */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    var @out = append(s, v);
    return @out;
}

internal static void setFirst<E>(slice<E> s, E v)
    where E : /* Integer */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    if (len(s) > 0) {
        s[0] = v;
    }
}

public static void PassSlice<S, E>(S s, E v)
    where S : /* ~[]E */ ISlice<E>, ISupportMake<S>, ISliceWrap<S, E>, new()
    where E : /* Integer */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    setFirst(new slice<E>(s), v);
}

[GoType("map[@string, nint]")] partial struct Grades;

public static bool EqualMaps<M, K, V>(M m1, M m2)
    where M : /* ~map[K]V */ IMap<K, V>, ISupportMake<M>, new()
    where K : /* comparable */ new()
    where V : /* comparable */ new()
{
    if (len(m1) != len(m2)) {
        return false;
    }
    foreach (var (k, v1) in m1) {
        {
            var (v2, ok) = m2[k, ꟷ]; if (!ok || !AreEqual(v1, v2)) {
                return false;
            }
        }
    }
    return true;
}

public delegate void Seq<V>(Func<V, bool> yield);

public delegate void KVSeq<K, V>(Func<K, V, bool> yield);

internal static Seq<nint> countdown(nint n) {
    return (Func<nint, bool> yield) => {
        for (nint i = n; i > 0; i--) {
            if (!yield(i)) {
                return;
            }
        }
    };
}

internal static KVSeq<@string, nint> letters() {
    return (Func<@string, nint, bool> yield) => {
        _ = yield("a"u8, 1) && yield("b"u8, 2);
    };
}

public static M CloneOrNil<M, K, V>(M m)
    where M : /* ~map[K]V */ IMap<K, V>, ISupportMake<M>, new()
    where K : /* comparable */ new()
    where V : /* comparable */ new()
{
    if (m.IsNil) {
        return default!;
    }
    var @out = make<M>(len(m));
    foreach (var (k, v) in m) {
        @out[k] = v;
    }
    return @out;
}

public static void DropKey<M, K, V>(M m, K key)
    where M : /* ~map[K]V */ IMap<K, V>, ISupportMake<M>, new()
    where K : /* comparable */ new()
    where V : /* comparable */ new()
{
    delete(m, key);
}

internal static void Main() {
    Point p = default!;
    p = new int32[]{1, 2, 3}.slice();
    ScaleAndPrint(p);
    fmt.Println(Twice(new Point(new int32[]{3, 4}.slice()), 2).String());
    var (dst2, src2) = (new Point(new int32[]{0, 0, 0}.slice()), new Point(new int32[]{7, 8, 9}.slice()));
    var (lo, hi) = CopyClearMinMax<Point, int32>(dst2, src2);
    fmt.Println(dst2.String(), src2.String(), lo, hi);
    var q = new Point(new int32[]{1, 2, 3, 4}.slice());
    var total = SumHalves<Point, int32>(q);
    fmt.Println(q.String(), total);
    var grown = AppendKeep(new Point(new int32[]{5, 6}.slice()), 7);
    fmt.Println(grown.String());
    var pt = new Point(new int32[]{1, 2}.slice());
    PassSlice(pt, 42);
    fmt.Println(pt.String());
    var g1 = new Grades(new map<@string, nint>{["a"u8] = 1, ["b"u8] = 2});
    var g2 = new Grades(new map<@string, nint>{["a"u8] = 1, ["b"u8] = 2});
    var g3 = new Grades(new map<@string, nint>{["a"u8] = 9});
    fmt.Println(EqualMaps<Grades, @string, nint>(g1, g2), EqualMaps<Grades, @string, nint>(g1, g3));
    nint sum = 0;
    foreach (var v in range<nint>(countdown(5).Invoke)) {
        if (v == 2) {
            break;
        }
        sum += v;
    }
    fmt.Println(sum);
    foreach (var (k, v) in range<@string, nint>(letters().Invoke)) {
        fmt.Println(k, v);
    }
    Grades nilG = default!;
    fmt.Println(CloneOrNil<Grades, @string, nint>(nilG) == default!);
    var cl = CloneOrNil<Grades, @string, nint>(g1);
    fmt.Println(len(cl), cl["a"u8]);
    DropKey<Grades, @string, nint>(g1, (@string)"a");
    fmt.Println(len(g1));
    var (i16, ok16) = bsearchLike(new uint16[]{2, 4, 8}.slice(), (uint16)4);
    var (i32, ok32) = bsearchLike(new uint32[]{10, 20, 30}.slice(), (uint32)25);
    fmt.Println(i16, ok16, i32, ok32, halve((uint16)6), halve((uint32)100));
    fmt.Println(halveN((int32)10), halveN((int64)(-3)));
    var ssum = (int64)0;
    foreach (var v in range<int64>(seqOf<int64, int32>((int32)4).Invoke)) {
        ssum += v;
    }
    fmt.Println(ssum);
    fmt.Println(growShrink((uintptr)1, (uintptr)20), growShrink((uint32)2, (uint32)9));
}

internal static U growShrink<U>(U v, U lim)
    where U : /* ~uint32 | ~uintptr */ IAdditionOperators<U, U, U>, ISubtractionOperators<U, U, U>, IMultiplyOperators<U, U, U>, IDivisionOperators<U, U, U>, IIncrementOperators<U>, IDecrementOperators<U>, IUnaryNegationOperators<U, U>, IModulusOperators<U, U, U>, IBitwiseOperators<U, U, U>, IShiftOperators<U, int, U>, IEqualityOperators<U, U, bool>, IComparisonOperators<U, U, bool>, new()
{
    while (v < lim) {
        v = v * ConvertToType<U>(2) + ConvertToType<U>(1);
    }
    return ((v % lim) >> (int)(1));
}

internal static Seq<T> seqOf<T, N>(N n)
    where T : /* ~int64 */ IAdditionOperators<T, T, T>, ISubtractionOperators<T, T, T>, IMultiplyOperators<T, T, T>, IDivisionOperators<T, T, T>, IIncrementOperators<T>, IDecrementOperators<T>, IUnaryNegationOperators<T, T>, IModulusOperators<T, T, T>, IBitwiseOperators<T, T, T>, IShiftOperators<T, int, T>, IEqualityOperators<T, T, bool>, IComparisonOperators<T, T, bool>, new()
    where N : /* ~int32 | ~int64 */ IAdditionOperators<N, N, N>, ISubtractionOperators<N, N, N>, IMultiplyOperators<N, N, N>, IDivisionOperators<N, N, N>, IIncrementOperators<N>, IDecrementOperators<N>, IUnaryNegationOperators<N, N>, IModulusOperators<N, N, N>, IBitwiseOperators<N, N, N>, IShiftOperators<N, int, N>, IEqualityOperators<N, N, bool>, IComparisonOperators<N, N, bool>, new()
{
    return (Func<T, bool> yield) => {
        for (var i = ConvertToType<T>(0); i < ConvertToType<T>(ConvertToUInt64<N>(n)); i++) {
            if (!yield(i)) {
                return;
            }
        }
    };
}

internal static (nint, bool) bsearchLike<S, E>(S s, E v)
    where S : /* ~[]E */ ISlice<E>, ISupportMake<S>, ISliceWrap<S, E>, new()
    where E : /* ~uint16 | ~uint32 */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    nint n = len(s);
    nint i = 0;
    nint j = n;
    while (i < j) {
        nint h = i + ((j - i) >> (int)(1));
        if (s[h] < v){
            i = h + 1;
        } else {
            j = h;
        }
    }
    return (i, i < n && AreEqual(s[i], v));
}

internal static Int halveN<Int>(Int n)
    where Int : /* ~int32 | ~int64 */ IAdditionOperators<Int, Int, Int>, ISubtractionOperators<Int, Int, Int>, IMultiplyOperators<Int, Int, Int>, IDivisionOperators<Int, Int, Int>, IIncrementOperators<Int>, IDecrementOperators<Int>, IUnaryNegationOperators<Int, Int>, IModulusOperators<Int, Int, Int>, IBitwiseOperators<Int, Int, Int>, IShiftOperators<Int, int, Int>, IEqualityOperators<Int, Int, bool>, IComparisonOperators<Int, Int, bool>, new()
{
    if (n <= ConvertToType<Int>(0)) {
        return n;
    }
    return ConvertToType<Int>(ConvertToUInt64<Int>(n) / 2);
}

internal static E halve<E>(E v)
    where E : /* ~uint16 | ~uint32 */ IAdditionOperators<E, E, E>, ISubtractionOperators<E, E, E>, IMultiplyOperators<E, E, E>, IDivisionOperators<E, E, E>, IIncrementOperators<E>, IDecrementOperators<E>, IUnaryNegationOperators<E, E>, IModulusOperators<E, E, E>, IBitwiseOperators<E, E, E>, IShiftOperators<E, int, E>, IEqualityOperators<E, E, bool>, IComparisonOperators<E, E, bool>, new()
{
    return (v >> (int)(1));
}

} // end main_package
